File: kotlin-typescript/karakum/src/jsMain/resources/plugins/convertCollections.js
import ts from "typescript";

export default function (node, context, render) {
    if (ts.isArrayTypeNode(node)) {
        return `js.array.ReadonlyArray<${render(node.elementType)}>`
    }
    if (
        ts.isIdentifier(node)
        && ts.isTypeReferenceNode(node.parent)
        && node.text === "Map"
    ) {
        return "js.collections.ReadonlyMap"
    }

    if (
        ts.isIdentifier(node)
        && ts.isTypeReferenceNode(node.parent)
        && node.text === "Set"
    ) {
        return "js.collections.ReadonlySet"
    }
    return null
}

File: kotlin-typescript/karakum/src/jsMain/kotlin/typescript/karakum/plugins/convertCollections.kt
package typescript.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.isIdentifier
import typescript.isTypeReferenceNode

val convertCollections = createPlugin { node, _, _ ->
    nullable {
        ensure(isIdentifier(node))
        ensure(node.text == "Map")

        val typeReference = ensureNotNull(node.getParentOrNull())
        ensure(isTypeReferenceNode(typeReference))

        "js.collections.ReadonlyMap"
    } ?: nullable {
        ensure(isIdentifier(node))
        ensure(node.text == "Set")

        val typeReference = ensureNotNull(node.getParentOrNull())
        ensure(isTypeReferenceNode(typeReference))

        "js.collections.ReadonlySet"
    }
}
